linkerd/app/outbound/src/http/peer_proxy_errors.rs

```diff
@@ -41,146 +37,89 @@ where
     type Error = S::Error;
     type Future = Pin<Box<dyn Future<Output = Result<S::Response, S::Error>> + Send + 'static>>;
 
+    #[inline]
     fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
         self.inner.poll_ready(cx)
     }
 
-    fn call(&mut self, req: Request<A>) -> Self::Future {
+    fn call(&mut self, req: http::Request<A>) -> Self::Future {
         let client = req.extensions().get::<ClientHandle>().cloned();
         debug_assert!(client.is_some(), "Missing client handle");
-        let response = self.inner.call(req);
-        Box::pin(async move {
-            let response = response.await?;
-            if let Some(message) = response.headers().get(L5D_PROXY_ERROR) {
-                tracing::info!(?message, "response contained `{}` header", L5D_PROXY_ERROR);
+
+        Box::pin(self.inner.call(req).map_ok(move |rsp| {
+            if let Some(msg) = rsp.headers().get(L5D_PROXY_ERROR) {
+                tracing::debug!(header = %L5D_PROXY_ERROR, ?msg);
+
                 // Gracefully teardown the accepted connection.
                 if let Some(ClientHandle { close, .. }) = client {
                     tracing::trace!("connection closed");
                     close.close();
                 }
             }
-            Ok(response)
-        })
+
+            rsp
+        }))
     }
 }
 
 #[cfg(test)]
 mod test {
-    use crate::{
-        http::{Endpoint, NewServeHttp, Request, Response, StatusCode, Version},
-        test_util::{
-            self, future,
-            support::{self, connect::Connect, http_util},
-        },
-        transport::addrs::{Remote, ServerAddr},
-        Config, Outbound,
-    };
-    use hyper::{client::conn::Builder, server::conn::Http, service, Body};
+    use super::*;
+    use crate::test_util;
+    use futures::future;
     use linkerd_app_core::{
-        errors::L5D_PROXY_ERROR,
-        io,
-        proxy::api_resolve::Metadata,
-        svc::NewService,
-        svc::{BoxNewService, BoxService},
-        tls::{ConditionalClientTls, NoClientTls},
-        Error, Infallible, ProxyRuntime,
+        svc::{self, ServiceExt},
+        Infallible,
     };
-    use linkerd_proxy_http::BoxRequest;
     use linkerd_tracing::test;
-    use std::net::SocketAddr;
+    use tokio::time;
 
     #[tokio::test(flavor = "current_thread")]
     async fn connection_closes_after_response_header() {
         let _trace = test::trace_init();
 
-        // Build the outbound server that responds with the l5d-proxy-error
-        // header set.
-        let (rt, _shutdown) = test_util::runtime();
-        let addr = SocketAddr::new([127, 0, 0, 1].into(), 12345);
-        let connect = support::connect().endpoint_fn_boxed(addr, |_: Endpoint| serve());
-        let config = test_util::default_config();
-        let mut outbound = build_outbound(config, rt, connect);
-
-        // Build the otubound service.
-        let service = outbound.new_service(Endpoint {
```

Comment:
One important advantage to the test changes is that this test won't have to be updated as the outbound stack (or target types) change.